### PR TITLE
feat(datetime): Add component

### PR DESCRIPTION
**Pull request by harel**
This is a work in progress mainly focused on exploring the possiblity of adding a datetime component which supports date ranges, time and localisation.

#### Bugs and Features
- [ ] allow typing/pasting dates into the input field
- [x] do not show completely disabled weeks in month view
- [ ] Menu needs some work, buttons don't make sense in all states
- [ ] Expose ways to customize the menu and table as well
- [ ] Add some more props for configuring the view and doc examples
  - [ ] showWeekends
  - [ ] showDaysOfWeek
- [ ] if at all possible, remove inline styles.  required styles should be configurable.
- [ ] move to 5 button CalendarMenu: `<<    <      Mo, Yr      >    >>`.  double chevrons change year, single change month, mo
- [ ] don't allow selecting invalid ranges: start > end
- [ ] use 12hr time grid with AM/PM option
- [ ] provide onChange callbacks with `(e, { value })` where `value` is the proposed new value.
- [ ] Add dateSeparator prop to DateRange to control the separator string between the two dates

#### Docs
- [ ] document function prop signatures
- [x] change `Week starts Sunday` example to `First Day Of Week` show two examples, start on Sun, another on Monday.
- [x] fix disabled days example, do not import date utils
#### Cleanup
- [x] move component to `/addons`
- [x] remove cruft propTypes that do not exist
- [ ] sort propTypes alphabetically
- [ ] update Month.js to use DateTimeGrid, https://github.com/Semantic-Org/Semantic-UI-React/pull/1240#issuecomment-289264758
- [ ] complete all code `TODO`s in `/Datetime`
- [x] reverse `onDateSelect` args to match other callbacks, with `event` first.
- [ ] remove unused methods in classes

### CODE REVIEW

File: src/modules/Datetime/CalendarMonth.js [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  const classes = cx(
+    className,
+  )
+  const rest = getUnhandledProps(CalendarMonth, props)
+  const ElementType = getElementType(CalendarMonth, props)
+
+  return <ElementType {...rest} className={classes}>{children}</ElementType>
+}
+
+DropdownMenu._meta = {
```
Comment (levithomason) [inline src/modules/Datetime/CalendarMonth.js]: CalendarMonth._meta

File: src/modules/Datetime/CalendarMonth.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+  name: 'CalendarMonth',
+  parent: 'Datetime',
+  type: META.TYPES.MODULE,
+}
+
+DropdownMenu.propTypes = {
```
Comment (levithomason) [inline src/modules/Datetime/CalendarMonth.js]: CalendarMonth.propTypes

**Review by levithomason — COMMENTED** [review]
I realize this is work in progress. Couple typos that might trip you up:

**Comment by harel** [comment]
That component wasn't imported yet but fixed the typos. Did not solve the problem.

image

**Comment by levithomason** [comment]
Thanks for the note, I'll take a look at this branch later tonight and resolve the issue.

**Comment by levithomason** [comment]
Fixed:

```diff
-static PropTypes
+static propTypes
```

I also pushed a lot of cleanup so I could browse the code a little easier.

**Comment by levithomason** [comment]
Looks like we'll need to reabse this branch to master to get the commits in order.  All the typings were moved from `index.d.ts` files to individual component files such as `Button.d.ts`.  There is still an `index.d.ts` file but it just exports the default component file from the directory.

This is 1:1 with how the JS files are setup as well.  LMK if you run into any issues with the rebase cleanup and I'll help out.

**Comment by harel** [comment]
No worries. I rebase periodically anyway so this will be just one more link in the chain. Didn't do much over the last few days because of life but i'm back at it this week.

**Comment by levithomason** [comment]
I've pushed some cleanup and a rebase to the latest master.

I think we should start breaking this PR down into small shipments and releasing it.  This will allow us to focus on one chunk at a time and do a good job on just that.  As we polish and release each piece, we can bring it back into this branch and prepare the next release.

Do you think we could somehow release Date only as the first component, the Time, and Datetime?  I'm wondering if it makes sense to break these down that way API wise as it does release wise.  Let me know your thoughts.

**Comment by levithomason** [comment]
Heads up, I'll be pushing some more refactors and cleanups for a bit here.

**Comment by levithomason** [comment]
OK, after pushing the latest I think we can ship Datetime with basic support first, then perhaps follow-up with the range support.

### DateTimeGrid

I pushed a generic DateTimeGrid that is used in all but the Month view.  It allows an easier construct the table cells, just pass a single array of as many cells as you want and tell the DateTimeGrid how many columns to fit them into.  It will chunk out the Rows for you.

This will also give us the ability to handle different layouts, since the DateTimeGrid just takes a single array of cell props objects.  We can use this array to build any layout.  We can even, later on, expose a `renderGrid` prop and allow the user to render their own calendar grids if we wanted to.

### TODO

I'd like to factor out the DayCell into the DateTimeGrid but am not 100% sure on how yet.  I'd also like to move the dateUtils to the Datetime directory and perhaps remove more, if possible (I removed some already).

***

That's all I got for now, thanks for all the work on this and the patience dealing with my limited availability.  We're getting close to something here 👍

**Comment by KevinGorjan** [comment]
Whats the status of this issue?

**Comment by GregoryPotdevin** [comment]
I'm also interested in this feature and can dedicate some time this week to test/code/document/brainstorm or whatever is needed.

**Comment by harel** [comment]
It's coming together, late stage development. There are a few things that
require feedback - in particular the momentjs branch which decouples date
handling to support both native js dates as well as moment/moment-timezone
dates (or any other time framework).

On Mon, Apr 24, 2017 at 10:19 AM, Gregory Potdevin <notifications@github.com

**Comment by viiiprock** [comment]
Hi guys, how is it going :)

**Comment by neslinesli93** [comment]
Any update on this?

**Comment by luizhenriquesoares** [comment]
good work guys!

**Comment by levithomason** [comment]
Update:

This branch has some failing conformance tests that need resolved.  Once those are resolved, I want to trim any non-functioning parts and ship it.

If anyone cares to dive in, please do!  Fork this branch and run `yarn test:watch` to see what is left to fix.

I continue to trickle work here as often as I can, which tends to be about once a week.

**Comment by levithomason** [comment]
OMG, this component is finally conformant!

image

We should be ready for some user testing here shortly :)

**Comment by revik** [comment]
Can't wait, any progress ? Do you need help testing it?

**Comment by dankirlin** [comment]
Let me know if there is anyway I can help push this forward.

**Comment by ydsood** [comment]
Waiting eagerly for this to be completed. This is a very frequently used component in forms. Available for helping push this out soon.

**Comment by neslinesli93** [comment]
Same here, count me in if you need anything

**Comment by levithomason** [comment]
If anyone would like to hop in here's some more info.  Conformance tests are passing and there are plenty of doc site examples.  What's needed is to run the doc site, try the examples, and fix the bugs.  Bug fixes should also have tests written for them.

As soon as all the doc site examples are working with passing tests, this can be shipped. 🎉

**Comment by srdjanRakic** [comment]
How is it going guys ? Any news on this ?

**Comment by Gallevy** [comment]
Patiently waiting. Anything I can do to speed this up?

**Comment by davidhenley** [comment]
This is the final piece missing for all my forms using SUI. Can't wait!

**Comment by hisuwh** [comment]
All I want for Christmas is SUI datepicker 🎄 🎁 📆 😆

**Comment by brianespinosa** [comment]
Hey everyone! I think we are all excited to ship this. At the same time, there is only so much bandwidth that main contributors to this project have to give to open source work. The nice thing is that it's open source and we are happy to accept pull requests from anyone that moves this project/feature forward.

[Per @levithomason's comment above](https://github.com/Semantic-Org/Semantic-UI-React/pull/1240#issuecomment-346949823), as soon as doc site examples are working and tests are passing for the reported bugs, this can be shipped. @Gallevy you asked if there is anything you can do to move this forward... we welcome any time you can dedicate to fixing any of the remaining bugs. Anyone else who has commented wanting to know what is going on, you are also welcome to help out. <3

**Comment by srdjanRakic** [comment]
@brianespinosa When will we be able to use this? :)

**Comment by therealgilles** [comment]
Could this PR be merged so that testing it is made easier? Thank you

**Comment by levithomason** [comment]
If you'd like to contribute to the development of this feature, please email me[at]levithomason.com.  This thread produces too much noise and distracts from the focus.

**Comment by levithomason** [comment]
A few collaborators have emailed me and pushed this work forward.  You are amazing ❤️.

I've ported this branch to our repo for easier management.  I'm merging any well authored PRs against this effort.

See #2821 for the Datetime branch on the main repo
See #2813 as an example of how you can help in making PRs against the feature branch

**Pull request closed without merging**